### ISSUE TITLE
Use `glean_client_info` to fill-in the client_info section

### ISSUE THREAD

**Issue by fbertsch** [mozilla/mozilla-schema-generator#159]
Currently, the client-info section is already defined in the [glean schema](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/glean/glean/glean.1.schema.json). However, there are platform-specified fields in there: e.g. `android_sdk_version`.

To accommodate this, we should find metrics with `send_in_pings: - glean_client_info` and fill those fields in to the `client_info` section.

**Comment by wlach** [comment]
@DawoodShahat -- can you look at this? It's not well specified right now (more instructions to come from @fbertsch), but maybe for a start you can poke around at this repo in addition to probe-scraper and glean (https://github.com/mozilla/glean). Feel free to ask questions in the channel if anything is unclear.

**Comment by DawoodShahat** [comment]
@fbertsch Hey, Can you elaborate on this a little bit more?  I'm new to the data-pipeline and trying to understand how mozilla-schema-generator and mozilla-pipeline-schemas fit together.

**Comment by fbertsch** [comment]
@DawoodShahat, sure! There's a good overview to start with here: https://docs.telemetry.mozilla.org/concepts/pipeline/schemas.html.

From there, you should be able to tell that this repo is doing two things:
1. Creating full schemas, based on the raw schemas in mozilla-pipeline-schemas and the probe data in probe-scraper
2. From those, generating BigQuery schemas and pushing those back to mozilla-pipeline-schemas

This task should include a few things:
- Updating the [glean config](https://github.com/mozilla/mozilla-schema-generator/blob/master/mozilla_schema_generator/configs/glean.yaml) to fill-in the client-info section of the ping with any probes that have `glean_client_info` as their `send_in_pings` value. (there may have to be some updates to the [glean ping handling](https://github.com/mozilla/mozilla-schema-generator/blob/master/mozilla_schema_generator/glean_ping.py)).
- A test of the glean_client_info section (https://github.com/mozilla/mozilla-schema-generator/blob/master/tests/test_glean.py)
- Verification that, when the entire mozilla-schema-generator is [run](https://github.com/mozilla/mozilla-schema-generator#development-and-testing), the BigQuery schemas do _not_ change.

**Comment by DawoodShahat** [comment]
I've been able to come up with how am I going to update the glean.yaml file to update client_info section of the probe that has ```glean_client_info``` as their ```send_in_pings```  value. 

```
client_info:
  android_sdk_version:
    history:
      match:
        send_in_pings:
          contains:
            - glean_client_info
  type: string
```

But I'm unable to understand how am I going to update the glean_ping.py to accommodate those changes.
Your guidance would be helpful, Thanks!

**Comment by acmiyaguchi** [comment]
It probably would be best to start with writing a failing test where you assert that certain values not already present in the `client_info` section but are in the `glean_client_info`.

https://github.com/mozilla/mozilla-schema-generator/blob/3cf5107800b45432b433fb7081c81a36c7617b99/tests/test_glean.py#L49

@travis79 provided some good context around what the `glean_client_info` tag means in the #glean-dictionary channel

> FYI, the glean_client_info is a special "ping_section" that is internal to Glean and included in every ping, information on what it contains is here: https://mozilla.github.io/glean/book/user/pings/index.html#the-client_info-section

> The metrics.yaml where the `glean_client_info` fields are defined is here: https://github.com/mozilla/glean/blob/main/glean-core/metrics.yaml


When you do get around to implementation, you'll want to update the `client_info` dictionary if there are metrics that match the `glean_client_info` tag. 

https://github.com/mozilla/mozilla-schema-generator/blob/3cf5107800b45432b433fb7081c81a36c7617b99/mozilla_schema_generator/glean_ping.py#L134-L161

**Comment by DawoodShahat** [comment]
Thanks, One question, how exactly am I going to get the client_info dictionary and update it?

**Comment by acmiyaguchi** [comment]
If you look at the schema values in the `super().generate_schema(new_config)` result dictionary, these are the actual schema schema objects before they're serialized to json.

https://github.com/mozilla/mozilla-schema-generator/blob/3cf5107800b45432b433fb7081c81a36c7617b99/mozilla_schema_generator/glean_ping.py#L152-L161

You would modify the `schema.schema["client_info"]` object.

If you read through the tests, you can see how you can access these schemas and play around with them. I suggest creating a new test, and running `pytest -k {name_of_your_test} --pdb` and inserting an `assert False` statement around where you want to debug. Alternatively, you can insert the following statements in the code to get an interpreter where you can play with the values using `import pdb; pdb.set_trace()`.

**Comment by DawoodShahat** [comment]
1. I wrote a config file to match client_info fields, e.g. android_sdk_version

```jsx
client_info:
  android_sdk_version:
    match:
      send_in_pings:
        contains: glean_client_info
      type: string
```

Matcher is configured with the above matcher_obj, and it matches the first probe from [probes](https://probeinfo.telemetry.mozilla.org/glean/glean/metrics)
 But it then throws an error, `Additional Properties cannot be missing for Glean probes.`

upon looking into it, I think it is because the client_info section in the [raw schema](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/glean/glean/glean.1.schema.json#L13-L19) doesn't have `additionalProperties` set in the same manner as they are set for the metrics section.
- **Metrics section**

```
 "boolean": {
       "additionalProperties": {
        "type": "boolean"
 },
```

- **client_info section**
```
"android_sdk_version": {
     "description": "The optional Android specific SDK version of the software running on this hardware device.",
      "type": "string"
}
```


I played with the test `test_single_schema()` It doesn't throw any error like the above and It has `additionalProperties`  set for each metric in the raw schema.

Some Queries
- What fields should I update the client_info with? As there are numerous fields in each probe that has `send_in_pings` set  to `glean_client_info`.
- We fill metrics section with two fields `description` and `name` of a probe whose `send_in_pings` isn't set to `glean_ping_info`. Isn't these two fields already set for `client_info` section in the raw schema?

@fbertsch Please, take a look at this whenever possible. Thanks!

**Comment by fbertsch** [comment]
>     1. I wrote a config file to match client_info fields, e.g. android_sdk_version
> 
> 
> ```js-jsx
> client_info:
>   android_sdk_version:
>     match:
>       send_in_pings:
>         contains: glean_client_info
>       type: string
> ```
> 
> Matcher is configured with the above matcher_obj, and it matches the first probe from [probes](https://probeinfo.telemetry.mozilla.org/glean/glean/metrics)
> But it then throws an error, `Additional Properties cannot be missing for Glean probes.`
> 
> upon looking into it, I think it is because the client_info section in the [raw schema](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/master/schemas/glean/glean/glean.1.schema.json#L13-L19) doesn't have `additionalProperties` set in the same manner as they are set for the metrics section.

Right, we are going to have to do a different route to get the schema for these metrics. Essentially we have to use the type to retrieve the appropriate schema. So the `string` type maps to `metrics/string`, which _will_ have an `additionalProperties` section.


>     * What fields should I update the client_info with? As there are numerous fields in each probe that has `send_in_pings` set  to `glean_client_info`.

All of these should be included in the `client_info` section.

>     * We fill metrics section with two fields `description` and `name` of a probe whose `send_in_pings` isn't set to `glean_ping_info`. Isn't these two fields already set for `client_info` section in the raw schema?

I'm not totally sure I'm following. We want to overwrite whatever is in the `client_info` section with the probes that have `glean_client_info` as the `send_in_pings` value. We _do_ want them to have the correct description and name.

**Comment by fbertsch** [comment]
cc @DawoodShahat^
